### PR TITLE
[clkmgr] Initial Clock Manager DIF

### PR DESCRIPTION
**Pull request by lenary**
This PR adds an initial Clock Manager DIF.

The clock manager is as yet undocumented, beyond the register file in `hw/top_earlgrey/ip/clkmgr/data/autogen/clkmgr.hjson` (note this is topgen'd, ignore other `clkmgr.hjson` files in the repo). Not all clocks are software-accessible, so not all can be controlled via the register interface in the clock manager.

The intention is to use the values of `top_earlgrey_gateable_clocks_t` in params of type `dif_clkmgr_gateable_clock_t`, and likewise for hintable clocks.

### CODE REVIEW

File: sw/device/lib/dif/dif_clkmgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_clkmgr.h:56]: Might be worth going through and checking the grammar on the comments. (The template instantiator is good but not perfect.)

File: sw/device/lib/dif/dif_clkmgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_clkmgr.h:139]: What do you think about
```
dif_clkmgr_gateable_get_enabled
dif_clkmgr_gateable_set_enabled
```
in analogy with the IRQ functions? (Same for hintable clocks; I dropped the `_clock` path component but that's just me.)

File: sw/device/lib/dif/dif_clkmgr.h
Comment (mcy) [inline sw/device/lib/dif/dif_clkmgr.h:192]: If this winds up being a pair of functions, do you think `dif_clkmgr_get_hint` and `set_hint` are more succinct?

**Review by mcy — COMMENTED** [review]
Just some thoughts on names.

File: sw/device/lib/dif/dif_clkmgr.h
Comment (lenary) [inline sw/device/lib/dif/dif_clkmgr.h:91]: I hoped there would be parameters which told me how many clocks were gateable/hintable, but there aren't, so it's not going to be that much fun to validate these indexes.

File: hw/top_earlgrey/sw/autogen/top_earlgrey.h
Comment (silvestrst) [inline hw/top_earlgrey/sw/autogen/top_earlgrey.h:812]: Missing enum variant descriptions.

File: hw/top_earlgrey/sw/autogen/top_earlgrey.h
Comment (silvestrst) [inline hw/top_earlgrey/sw/autogen/top_earlgrey.h:825]: Same as above.

File: util/topgen/c.py
Comment (silvestrst) [inline util/topgen/c.py:400]: Function description would be useful.

**Review by silvestrst — COMMENTED** [review]
`[topgen] Generate Clock Description for SW` commit review.

- Please add commit message describing why this change is necessary and what issues it solves.

- Please make this change compile, so that reviewers can generate and have a look at the `clkmgr_regs.h` header file.

File: sw/device/lib/dif/dif_clkmgr.h
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.h:124]: `dif_clkmgr_gateable_get_enabled` see [@mcy comment](https://github.com/lowRISC/opentitan/pull/3853/commits/6ab31ecd349f58b389bab0cc308c0bd08bbd9ee0#r508558029)

File: sw/device/lib/dif/dif_clkmgr.h
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.h:156]: As above `get_enabled`?

File: sw/device/lib/dif/dif_clkmgr.h
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.h:183]: This explanation I find confusing. This function allows you to find out what hint is set for the clock, but `accurately reflects the state of the current clock` but this is almost contradictory.

File: sw/device/lib/dif/dif_clkmgr.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_clkmgr.h:192]: I like @mcy suggestion!

**Review by silvestrst — COMMENTED** [review]
@lenary thanks, looks good!

Some minor comments, and thumbs up to @mcy suggestions.

File: sw/device/lib/dif/dif_clkmgr.h
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.h:99]: Is this completely correct?

The documentation says the following:
```
        Clock hint for software gateable clocks.
        These clocks are not fully controlled by software.

        For disable, software only provides a hint, and hardware determines the final clock state based on the
        hint and whether the block in question is idle.
```

Which only describes a special restriction in regard to the `disable` operation, says nothing about `enable` operation.

The individual bits description also says the following:
```
      fields: [
        {
          bits: "0",
          name: "CLK_MAIN_AES_HINT",
          resval: 1,
          desc: '''
            0 CLK_MAIN_AES can be disabled.
            1 CLK_MAIN_AES is enabled.
          '''
        }
```
Which  only for the `disable` has `can` in the description. The `enable` description is the same as for the gateable clocks.

File: sw/device/lib/dif/dif_clkmgr.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_clkmgr.h:183]: Am I understanding correctly that this function is actually reading `CLK_HINTS_STATUS`, which gives the "real" state of the clock? I think it should be named `get_status` or `get_state` to avoid confusion.

**Review by gkelly — COMMENTED** [review]
This hasn't gone through design review, that's happening tomorrow (@silvestrst and I) so hold off on reviews until this is complete please.

**Comment by silvestrst** [comment]
> This hasn't gone through design review, that's happening tomorrow (@silvestrst and I) so hold off on reviews until this is complete please.

I think the better approach here is probably creating a DRAFT PR with only the "first header reviewers" added.

File: sw/device/lib/dif/dif_clkmgr.h
Comment (gkelly) [reply sw/device/lib/dif/dif_clkmgr.h:183]: We agree with @silvestrst, if the function reads the actual state of the clock then the name should reflect that it's reading the state not the hint. This looks like it's maybe just a copy and paste from the description above, so perhaps unintentional?

File: sw/device/lib/dif/dif_clkmgr.h
Comment (gkelly) [reply sw/device/lib/dif/dif_clkmgr.h:192]: `set`/`get` are idiomatic with our other DIFs, so these functions should become `hintable_clock_get_enabled`, `hintable_clock_set_hint`, `hintable_clock_get_hint`.

File: sw/device/lib/dif/dif_clkmgr.h
Comment (gkelly) [reply sw/device/lib/dif/dif_clkmgr.h:91]: We're not sure what you mean here, can you clarify? The generated enums have a `Last` variant, so isn't this just checking that the value is `< Last` and returning `BadArg` if not?

**Review by gkelly — APPROVED** [review]
Thanks for doing this header @lenary! We (@silvestrst and I) have done a header review. We are broadly happy with this header, with the following comments to address before submission:
- It would be nice if this built (add an empty `.c` and compile target), just as a smoke test. It's a little easier to look at the generated regs when writing the `.c` if there's a building target. 
- The naming of the accessors that look at the state of gated and hinted clocks should be updated to be inline with the comments we've left, just to make them idiomatic.

Thanks again for the DIF header. We believe the header is good to submit once you've addressed the comments.

Thanks!

File: sw/device/lib/dif/dif_clkmgr.h
Comment (lenary) [reply sw/device/lib/dif/dif_clkmgr.h:91]: In other DIFs, we normally check that the value of an identifier like this is in the right range (which is normally less than uint32_t) - the common way of doing this in guard clauses is to check against definitions provided by the reggen'd file, for instance `dif_plic_target_t` is always checked to be less than `RV_PLIC_PARAM_NUMTARGET` in `dif_plic.h` - the problem here is there are no parameters from the clockmanager's hjson file that we can check this index against.

We don't use the topgen `Last` definition, because that mixes dif-code with top-specific code, something we aim not to do.

**Review by silvestrst — APPROVED** [review]

File: sw/device/lib/dif/dif_clkmgr.h
Comment (lenary) [reply sw/device/lib/dif/dif_clkmgr.h]: Yeah, this was a copy/paste that was not updated. New naming and description should be clearer.

File: sw/device/lib/dif/dif_clkmgr.h
Comment (silvestrst) [reply sw/device/lib/dif/dif_clkmgr.h:90]: We probably could pass these values in the `init` function, I think maybe timer and alert manager does something similar?

Comment (lenary) [reply sw/device/lib/dif/dif_clkmgr.h:90]: I've filed #3932 so it's done the same way as the PLIC is, for the moment.

**Comment by lenary** [comment]
Current Status: DIF Functions implemented, but not tested. I'm starting on unit tests right now.

**Comment by lenary** [comment]
I discussed this with @silvestrst earlier this week - I've started on unit tests, but we'll do as we've done with other DIF PRs, and add them in a follow-up PR. 

This is ready for review.

File: sw/device/lib/dif/dif_clkmgr.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.c:15]: Are you sure you are using this define correctly? This just tells you how wide the registers are. I am not sure I understand how these are relatable to `kMaxGateableClocks` and `kMaxHintableClocks`.

I think `kMaxGateableClocks` and `kMaxHintableClocks` should be passed into the DIF as `dif_clkmgr_params_t params` to the `init` function (with values taken from the relevant `top_earlgrey` definitions).

File: sw/device/lib/dif/dif_clkmgr.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.c:21]: Same as above.

File: sw/device/lib/dif/dif_clkmgr.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.c:36]: `clock >= kMaxGateableClocks`

Even considering that `kMaxGateableClocks = CLKMGR_PARAM_REG_WIDTH` is correct, this check is of limited use as it won't capture all of the illegal values, as the description says:

```
/**
 * There are fewer Hintable clocks than this in the hardware, but this is the
 * threshold at which we will need to change the implementation of the DIF.
 */
```
Which implies that there are fewer legal clock values than covered by this constant.

Comment (silvestrst) [reply sw/device/lib/dif/dif_clkmgr.c:36]: This also applies in the other functions where this check is performed, and also the same with hintable clock functions.

File: sw/device/lib/dif/dif_clkmgr.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.c:63]: Not sure whether `kDifClkmgrBadArg` or `kDifClkmgrError` is better suited here. I think `kDifClkmgrBadArg` is probably the right error code as it would cover both variants (even if the value is glitched, it still is a bad arg).

File: sw/device/lib/dif/dif_clkmgr.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_clkmgr.c:133]: Forgot to delete this line?

**Review by silvestrst — COMMENTED** [review]
@lenary thanks, it is looking good!

Mostly small comments, I think comment regarding the `kMaxGateableClocks = CLKMGR_PARAM_REG_WIDTH` is potentially only real issue?

Also, linter wants you to tun `make -C hw top`:
`##[error]Autogenerated tops not up-to-date. Regenerate with 'make -C hw top'.`

File: sw/device/lib/dif/dif_clkmgr.c
Comment (lenary) [reply sw/device/lib/dif/dif_clkmgr.c:15]: We're waiting on https://github.com/lowRISC/opentitan/issues/3932 to get these values at all - until then, assuming 32 is not bad, as we know we maximum have one register of them. 

In the long run we probably do want these in `params_t`, but so far we haven't required that DIFs are fully parameterisable, just that we want to get there eventually.

File: sw/device/lib/dif/dif_clkmgr.c
Comment (lenary) [reply sw/device/lib/dif/dif_clkmgr.c:133]: Thanks, will delete.

File: sw/device/lib/dif/dif_clkmgr.c
Comment (lenary) [reply sw/device/lib/dif/dif_clkmgr.c:62]: BadArg is allowed here, as we haven't yet accessed an MMIO - I could see a different implementation requiring Error here.

**Comment by lenary** [comment]
I have opened https://github.com/lowRISC/opentitan/issues/4201 to track follow-ups, specifically including supporting more than 32 clocks.

File: sw/device/lib/dif/dif_clkmgr.c
Comment (lenary) [reply sw/device/lib/dif/dif_clkmgr.c]: I took the other approach of adding `last_hintable_clock` and `last_gateable_clock` to `dif_clkmgr_params_t`.

**Review by silvestrst — APPROVED** [review]
@lenary thanks!

Looks Good To Me! Thank you for addressing the comments!

**Pull request merged**
